<a href="https://colab.research.google.com/github/MariaDaguer/Analitica-de-negocios/blob/main/Tech_Sales_Report_Reto2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Pre-operación

In [ ]:
import numpy as np
import pandas as pd


from  sklearn.tree import  DecisionTreeClassifier
from sklearn.metrics import confusion_matrix




1. Base de datos y variable

In [ ]:
nxl='/content/2. TechSalesReport_Int_L.xlsx'
XDB=pd.read_excel(nxl, sheet_name=2)

XDB.head(50)

XD= XDB[['Business','Age','Female','Years','Certificates','Feedback','Salary','NPS']]
yd=XDB[['College']]


Sales_Rep  Business  Age  Female  Years  College  Certificates  Feedback  \
0           1         0   59       1      2        1             1      2.01   
1           2         0   52       0     10        1             4      3.64   
2           3         1   47       1      1        1             1      3.88   
3           4         0   61       0      2        1             3      2.70   
4           5         1   39       0      1        0             5      3.44   
5           6         0   28       0      6        1             1      2.43   
6           7         1   25       1      1        1             5      3.30   
7           8         0   51       1     10        0             0      2.15   
8           9         0   34       0      4        1             2      2.91   
9          10         0   38       1      1        1             5      1.23   
10         11         1   53       1     11        1             2      3.93   
11         12         0   41       1      1        1             1      2.26   
12         13         0   40       0      1        0             4      3.60   
13         14         1   41       0      2        1             3      2.17   
14         15         0   46       1      2        1             1      4.00   
15         16         0   38       1      4        1             1      2.37   
16         17         1   39       0      2        0             2      3.00   
17         18         0   52       1      1        0             0      2.72   
18         19         1   54       1      2        1             5      2.04   
19         20         0   24       0      1        1             1      2.66   
20         21         1   53       0      1        1             2      2.42   
21         22         0   37       1      1        1             2      3.71   
22         23         1   32       1      1        1             3      3.42   
23         24         0   33       0      2        1             1      1.16   
24         25         1   55       1      1        1             1      1.99   
25         26         1   27       1      2        1             6      2.93   
26         27         0   58       1      2        1             4      3.54   
27         28         1   52       1      7        1             1      3.37   
28         29         1   26       0      1        1             2      3.29   
29         30         1   28       1      8        1             3      2.50   
30         31         0   62       0      7        1             5      2.60   
31         32         1   44       0      2        1             3      1.60   
32         33         0   42       1      2        1             5      2.15   
33         34         1   36       1      1        0             5      2.11   
34         35         1   52       1      1        0             3      1.92   
35         36         1   53       0      2        0             1      3.08   
36         37         1   56       1      2        1             6      1.43   
37         38         1   51       1      2        1             3      3.63   
38         39         0   40       0      2        1             5      3.64   
39         40         1   49       0      2        0             1      2.06   
40         41         0   43       0      2        1             3      1.63   
41         42         1   40       0      2        1             5      2.67   
42         43         1   43       0      4        1             3      3.62   
43         44         0   59       0      6        1             1      4.00   
44         45         1   22       0      1        1             3      1.92   
45         46         1   48       0      2        1             0      1.52   
46         47         1   27       0      2        0             4      3.72   
47         48         0   38       0      3        0             2      1.85   
48         49         0   65       1      1        0             2      2.84   
49  

2. Implementación del modelo

In [ ]:
mar=DecisionTreeClassifier(criterion='gini', max_depth=8) #aquí le dejimos: trabaja con el indice de gini y con las 4 variables
mar.fit(XD,yd)
#mar= modelo de arbol
#max_depth: variables disponibles, profundidad (trabajar con este o menor número de varibales )
#gini: es el criterio que indica el grado de error del modelo en la clasificación de solicitantes

#SE PROCEDE CON EL DESPLIEGUE DEL ÁRBOL

from sklearn.tree import export_graphviz #exportar
from pydotplus import graph_from_dot_data #crear el gráfico


vs= ['Business','Age','Female','Years','Certificates','Feedback','Salary','NPS']

dot_data=export_graphviz(mar, feature_names=vs) #esto es para construir el arbol
graph=graph_from_dot_data(dot_data)
graph.write_png('1.Arbolcollege.png')


True

3. Medir el desempeño

In [ ]:
ydp=mar.predict(XD)
cm=confusion_matrix(yd, ydp)
print(cm)

VN=cm[0,0];FP=cm[0,1]; FN=[1,0]; VP=[1,1]


Exactitud=(VP+VN)/(VP+VN+FP+FN) # funcionamiento general del modelo
print( 'La Exactitud es',Exactitud)

Sensibilidad= VP/(VN+FN)
print('La Sensibilidad es',Sensibilidad)

Especificidad=VN/(VN+FP)  # como se comportó el modelo con los negados
print('La Especificidad es',Especificidad)

Precisión=VP/(VP+FP) # cual es el acierto clasificando los positivos
print('La Precisión es',Precisión)


TasaError=(FP+FN)/(VP+VN+FP+FN) #equivocación general
print('La Tasa de Error es', TasaError)

PredNC=(VN+FN)/(VN+FP+FN+VP) #como funciona el modelo negando créditos
print('La tasa de predicción negativa es',PredNC)




[[  671  3799]
 [  373 17147]]
La Exactitud es [0.15026834 0.15030195]
La Sensibilidad es [0.0014881  0.00149031]
La Especificidad es 0.15011185682326622
La Precisión es [0.00026316 0.00026316]
La Tasa de Error es [0.84973166 0.84969805]
La tasa de predicción negativa es [0.15026834 0.15007828]


In [ ]:
XDB2=pd.read_excel(nxl, sheet_name=2)
XDB2.head(100)

yd2=XDB2[['College']]
yd2.head(10)

#Donde están los que no poseen información
filas=np.where((np.isnan(yd2))==True)[0]
filas
print('las filas que no poseen información son')
print(filas)



#Predicción

XD2 = XDB2[['Business', 'Age', 'Female', 'Years', 'Certificates', 'Feedback', 'Salary', 'NPS']]

ydp2 = mar.predict(XD2)
print("Las predicciones son", ydp2)



#Informe
dfp = pd.DataFrame(np.column_stack((XD2, ydp2)))
dfp.columns = ['Business', 'Age', 'Female', 'Years', 'Certificates', 'Feedback', 'Salary', 'NPS', 'Predicción']

print(dfp.head())


dfp.to_excel('InformeCollege.xlsx', index=False)



print('Los negados pronosticados son:', len(np.where(ydp2 == 0)[0]))
print('Los aprobados pronosticados son:', len(np.where(ydp2 == 1)[0]))

las filas que no poseen información son
[]
Las predicciones son [1 1 1 ... 1 1 1]
   Business   Age  Female  Years  Certificates  Feedback    Salary   NPS  \
0       0.0  59.0     1.0    2.0           1.0      2.01   70200.0   5.0   
1       0.0  52.0     0.0   10.0           4.0      3.64  133000.0  10.0   
2       1.0  47.0     1.0    1.0           1.0      3.88   52600.0   8.0   
3       0.0  61.0     0.0    2.0           3.0      2.70   96000.0   6.0   
4       1.0  39.0     0.0    1.0           5.0      3.44  122000.0   7.0   

   Predicción  
0         1.0  
1         1.0  
2         1.0  
3         1.0  
4         1.0  
Los negados pronosticados son: 1044
Los aprobados pronosticados son: 20946


Análisis

Para la implementación de este modelo de árbol de decisión, se tomaron un total de 9 variables. 8 variables de entrada en las que se destacan: Negocio, Edad, Género femenino, Años, Certificados, Retroalimentación, Salario y NPS, y una variable de salida que permite evaluar si un empleado posee o no un titulo universitario.
Para la construcción del modelo se tuvieron un total de 21990 datos divididos en 17520 que fueron a la universidad y un total de 4470 que no tienen titulo universitario.

Luego de la ejecución del modelo se realizó una medición de desempeño que mostró resultados muy bajos en todas las métricas importantes, en el caso de la exactitud y la especificidad, solo se logró un 15% (los valores más altos, mientras que la sensibilidad y la precisión arrojaron resultados cerca del 1%. La predicción negativa también resultó en un porcentaje demasiado bajo, por lo que no resultó una sorpresa que para este modelo, la tasa de error sobrepasara el 80%.

Luego de la construcción del modelo se procedión con el pronóstico de la obtención de titulos universitarios para los empleados, del que más tarde se logró un informe para estos trabajadores en donde un total de 1044 no poseen titulo universitario y un total de 20946, si. De esa forma, se puede concluir que esta es una organización donde el paso por la universidad tiene más peso que otro tipo de certificaciones.